service/history/task/cross_cluster_task.go

```diff
@@ -208,47 +265,396 @@ func newCrossClusterTaskBase(
+//   for polling again
+// If the returned request is not nil
+// - the request can be returned to the target cluster
+func (t *crossClusterSourceTask) GetCrossClusterRequest() (request *types.CrossClusterTaskRequest, retError error) {
+	t.Lock()
+	defer func() {
+		if retError == nil && request == nil {
+			t.state = ctask.TaskStateAcked
+		}
+		t.Unlock()
+	}()
+
+	if !t.isValidLocked() {
+		return nil, errors.New("task invalidated")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), taskDefaultTimeout)
+	defer cancel()
+
+	taskInfo := t.GetInfo().(*persistence.CrossClusterTaskInfo)
+	_, mutableState, release, err := loadWorkflowForCrossClusterTask(ctx, t.executionCache, taskInfo, t.shard.GetMetricsClient(), t.logger)
+	if err != nil || mutableState == nil {
+		return nil, err
+	}
+	defer func() { release(retError) }()
+
+	request = &types.CrossClusterTaskRequest{
+		TaskInfo: &types.CrossClusterTaskInfo{
+			DomainID:            t.GetDomainID(),
+			WorkflowID:          t.GetWorkflowID(),
+			RunID:               t.GetRunID(),
+			TaskID:              t.GetTaskID(),
+			VisibilityTimestamp: common.Int64Ptr(t.GetVisibilityTimestamp().UnixNano()),
+		},
+	}
+
+	switch t.GetTaskType() {
+	case persistence.CrossClusterTaskTypeStartChildExecution:
+		attributes, taskState, err := t.getRequestForStartChildExecution(ctx, taskInfo, mutableState)
+		if err != nil || attributes == nil {
+			return nil, err
+		}
+		request.TaskInfo.TaskType = types.CrossClusterTaskTypeStartChildExecution.Ptr()
+		request.TaskInfo.TaskState = int16(taskState)
+		request.StartChildExecutionAttributes = attributes
+	case persistence.CrossClusterTaskTypeCancelExecution:
+		attributes, taskState, err := t.getRequestForCancelExecution(ctx, taskInfo, mutableState)
+		if err != nil || attributes == nil {
+			return nil, err
+		}
+		request.TaskInfo.TaskType = types.CrossClusterTaskTypeCancelExecution.Ptr()
+		request.TaskInfo.TaskState = int16(taskState)
+		request.CancelExecutionAttributes = attributes
+	case persistence.CrossClusterTaskTypeSignalExecution:
+		attributes, taskState, err := t.getRequestForSignalExecution(ctx, taskInfo, mutableState)
+		if err != nil || attributes == nil {
+			return nil, err
+		}
+		request.TaskInfo.TaskType = types.CrossClusterTaskTypeSignalExecution.Ptr()
+		request.TaskInfo.TaskState = int16(taskState)
+		request.SignalExecutionAttributes = attributes
+	default:
+		return nil, errUnknownCrossClusterTask
+	}
+
+	return request, nil
+}
+
+func (t *crossClusterSourceTask) getRequestForStartChildExecution(
+	ctx context.Context,
+	taskInfo *persistence.CrossClusterTaskInfo,
+	mutableState execution.MutableState,
+) (*types.CrossClusterStartChildExecutionRequestAttributes, processingState, error) {
+	initiatedEventID := taskInfo.ScheduleID
+	childInfo, ok := mutableState.GetChildExecutionInfo(initiatedEventID)
+	if !ok {
+		return nil, t.processingState, nil
```

Comment:
are you intentionally returning nil error here? Is it expected? Same for the below ones.